Concentus.Oggfile/OpusOggReadStream.cs

```diff
@@ -128,37 +125,82 @@ private bool Initialize()
         {
             try
             {
-                OggContainerReader reader = new OggContainerReader(_inputStream, true);
-                if (!reader.Init())
+                var oggContainerReader = new OggContainerReader(_stream, true);
+                if (!oggContainerReader.Init())
                 {
-                    _lastError = "Could not initialize stream";
+                    LastError = "Could not initialize stream";
                     return false;
                 }
 
-                //if (!reader.FindNextStream())
-                //{
-                //    _lastError = "Could not find elementary stream";
-                //    return false;
-                //}
-                if (reader.StreamSerials.Length == 0)
+                if (oggContainerReader.StreamSerials.Length == 0)
                 {
-                    _lastError = "Initialization failed: No elementary streams found in input file";
+                    LastError = "Initialization failed: No elementary streams found in input file";
                     return false;
                 }
 
-                int streamSerial = reader.StreamSerials[0];
-                _packetProvider = reader.GetStream(streamSerial);
+                int firstStreamSerial = oggContainerReader.StreamSerials[0];
+                _packetProvider = oggContainerReader.GetStream(firstStreamSerial);
+
+                GranuleCount = _packetProvider.GetGranuleCount();
+                PageCount = _packetProvider.GetTotalPageCount();
+
                 QueueNextPacket();
 
                 return true;
             }
             catch (Exception e)
             {
-                _lastError = "Unknown initialization error: " + e.Message;
+                LastError = "Unknown initialization error: " + e.Message;
                 return false;
             }
         }
 
+        /// <summary>
+        /// Seeks the Opus stream for a valid packet at the specified offset.
+        /// </summary>
+        /// <param name="offset">The sample offset.</param>
+        public void SeekTo(long offset)
+        {
+            // Find a packet based on offset and return 1 in the callback if the packet is valid
+            var foundPacket = _packetProvider.FindPacket(offset, GetPacketLength);
+
+            // Check of the found packet is valid
+            if (foundPacket == null || foundPacket.IsEndOfStream)
+            {
+                _endOfStream = true;
+                _nextDataPacket = null;
+                return;
+            }
+
+            // Just seek to this found packet and get the previous packet (preRoll = 1)
+            _packetProvider.SeekToPacket(foundPacket, 1);
```

Comment:
If the stream does not support seeking (e.g. it is an HTTP download stream or something) then what is the expected error handler? Do you allow seeking forwards but not backwards, or do you fail outright?

Reply:
Added checks, for now check the _stream to see if it's seekable.
HTTP Download streaming : I dont think that will work for any decoder I use in my WebDAV-AudioPlayer application, so I just read all the data into a memorystream, which allows all I need...

Reply:
I think that the SeekTo method should check if the stream is seekable first, and if not, throw an InvalidOperationException

Reply:
I did already add those checks.

However....

When calling this code in the **Initialize()** method:
``` c#
GranuleCount = _packetProvider.GetGranuleCount();
PageCount = _packetProvider.GetTotalPageCount();
```

The code does already do a seek to to end to find all garnules / pages?
So if the stream is not seekable, this will throw I think...

So maybe not to worry too much about CanSeek now?

Reply:
BTW : **pagecount** is not really used internally + externally, so better remove it, or keep it for future?

Reply:
Um......you're the one adding the `PageCount` code....?
On one hand I see the use case for the code to just know in advance the total length and page count of the file so you can report it on the UI of a media player. It would be quite handy to have.
On the other hand, it's not unreasonable for this code to be used for a continuous stream (voice chat, internet radio, whatever) where the stream length is infinite and non-seekable, so I would like a design that doesn't break those users.

I think the compromise might be to have a utility function to retrieve the total playback time of the stream, and users can call that function if they want to and they know that the underlying stream is a fixed-length file, but if the stream is infinite length or not seekable, then it just throws an InvalidOperationException

Reply:
Update the code, please take a look if this is ok.